[wip] src, inspector: support opted-in VM contexts

TimothyGu opened the pull request:
Works, doesn't crash<del>, but no tests or docs at this moment</del>.

V8's console is injected unconditionally upon every `attachContext` call, which may be not desirable.

Uses an approach inspired by @jgoz in https://github.com/nodejs/node/issues/7593#issuecomment-310128400.

/cc @eugeneo @bmeck

Fixes: nodejs#7593
Refs: nodejs#9272

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [ ] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
inspector

jgoz commented:
This is pretty close to a solution I started a couple days ago, but mine avoids refactoring Contextify, instead introducing a `ContextFromSandbox` function that is used by the inspector agent to extract a V8 context.

Anyway, I like your solution better @TimothyGu.

jgoz commented:
@timothygu I'll push my branch and you can lift the test I wrote for it if you like. Same API.

refack commented:
Ref: #13295
Why can't it be streamlined into `vm.runInDebugContext` or `vm.runInInspectorContext`?

TimothyGu commented:
@refack Debug context is going to be removed by the end of this year. I'm not sure what you are referring to by "Inspector context".

TimothyGu commented:
@jgoz Thanks, I'll have to look into how the inspector harness works, and your script is a good place to start!

refack commented:
> @refack Debug context is going to be removed by the end of this year. I'm not sure what you are referring to by "Inspector context".

1. One could argue that instead of removing `vm.runInDebugContext` we could fix/reimplement based on the mechanism in this PR.
2. Alternatively you could create the Inspector variant and name it `vm.runInInspectorContext` and it will be a wrapper that encapsulates the logic needed.

More specifically my question is does the two new endpoint have any use outside of `vm` associated operations? If not maybe they should be `internal/` and published only as `vm.context` methods?
```diff
module.exports = {
    open: (port, host, wait) => open(port, host, !!wait),
    close: process._debugEnd,
    url: url,		    url: url,
 +  Session,
 +  attachContext,
 +  detachContext
  };
```
Which leads me to the (trivial) observation that this PR would be easier to review if you provided a draft doc and/or some tests... 🤷‍♂️

jgoz commented:
@refack 

> 1. One could argue that instead of removing vm.runInDebugContext we could fix/reimplement based on the mechanism in this PR.
> 2. Alternatively you could create the Inspector variant and name it vm.runInInspectorContext and it will be a wrapper that encapsulates the logic needed.
>
> More specifically my question is does the two new endpoint have any use outside of vm associated operations? If not maybe they should be internal/ and published only as vm.context methods?

Unless there is another way to create V8 contexts in Node, I think you are correct that these are only useful for `vm` operations. My original reasoning for adding to the `inspector` API, rather than `vm` was as follows:

1. Avoid coupling `vm` to `inspector`/current debugging subsystem (and similarly, avoid future deprecations such as with `runInDebugContext`)
2. `inspector` API is currently designated "unstable", so if a better solution becomes possible (e.g., via support from the V8 team), deprecating this API will not be as painful

Something worth testing would be calling `attach` at the start of `runInContext` and then `detach` at the end. If this allowed debugging during execution of a script while allowing proper GC cleanup of contexts as well as re-use of contexts, it might be a better solution (possibly controlled with an opt-in or opt-out flag).

refack commented:
> 1. Avoid coupling `vm` to `inspector`/current debugging subsystem (and similarly, avoid future deprecations such as with `runInDebugContext`)
> 2. `inspector` API is currently designated "unstable", so if a better solution becomes possible (e.g., via support from the V8 team), deprecating this API will not be as painful
>
> Something worth testing would be calling `attach` at the start of `runInContext` and then `detach` at the end. If this allowed debugging during execution of a script while allowing proper GC cleanup of contexts as well as re-use of contexts, it might be a better solution (possibly controlled with an opt-in or opt-out flag).

A dilemma indeed...

I have some weird ideas like adding an implementation modules to `runInDebugContext` so it'll be:
```js
runInDebugContext(code, require('inspector'));
```

Or adding a scope method to `inspector` so:
```js
require('inspector').runInScope(context);
```

TimothyGu commented:
@jgoz That would actually work, but only synchronously. That's still a better default behavior though, so I implemented it and added a big caveat in the documentation.

I've added tests and docs, but still need to figure out the `console` issue.

----

As I was talking to @addaleax at JSConf China, when `contextCreated` is called, V8 [**unconditionally**](https://github.com/nodejs/node/blob/e0340af4551717ada908dc9fa1a4cd706004d728/deps/v8/src/inspector/inspected-context.cc#L36-L39) sets the sandbox's `console` property to its internal console. This is possibly undesirable for most users of the VM module.

On our side I foresee three options:

- `ignore` Do not make any changes to the sandbox
- `merge` Merge the methods so that inspector and original methods are both called
- `replace` Replace whatever `console` avaiable in the sandbox with the Inspector version

All three are implementable right now, but the first two are difficult to implement w/o side effects. I would prefer it if V8 adds a flag to disable this feature. I'm working on writing a CL to V8, but Google's custom tools require a bit of a setup.

TimothyGu commented:
Unfortunately, the `console` issue is complicated by https://github.com/v8/v8/commit/54271c21e2860273abd6018c398b04f672ef2615. 

----

To @ak239, @hashseed, and the rest of @nodejs/v8:

We are trying to add Inspector support for contexts created through Node.js's [VM](https://nodejs.org/api/vm.html) module. The way we do that can be roughly illustrated as:

```js
const inspector = require('inspector');
const vm = require('vm');
const sandbox = {
  customProp: 42
};
vm.createContext(sandbox);
inspector.attachContext(sandbox);
vm.runInContext('newProp = customProp + 1;', sandbox);
console.log(sandbox);
// Prints { customProp: 42, newProp: 43 }
```

Now, V8's `contextCreated` callback unconditionally initializes the global object's `console` property to an Inspector-aware version, but this new `console` property would either be added to `sandbox` even though the user never requested it to be added, or worse, overwrite an existing `sandbox.console` before the `inspector.attachContext` call.

This is not really a problem for the main (top) context, as we already provide a global `console` variable anyway, and we can just merge our console methods with V8's. But it is a problem for VM contexts, whose properties can always be manipulated outside the context.

In the V8 version we currently have, the `console` setting is [done](https://github.com/nodejs/node/blob/e0340af4551717ada908dc9fa1a4cd706004d728/deps/v8/src/inspector/inspected-context.cc#L36-L39) by a `context->Global()->Set`, so it is reasonably easy to revert that change just for VM. But after https://github.com/v8/v8/commit/54271c21e2860273abd6018c398b04f672ef2615, it is practically impossible to restore the original state.

Please offer some ideas about how we can and should proceed re. this issue. Thanks!

refack commented:
> In the V8 version we currently have, the console setting is done by a context->Global()->Set, so it is reasonably easy to revert that change just for VM. But after v8/v8@54271c2, it is practically impossible to restore the original state.
> 
> Please offer some ideas about how we can and should proceed re. this issue. Thanks!

The global assignment line line from https://github.com/v8/v8/commit/54271c21e2860273abd6018c398b04f672ef2615:
```c++
JSObject::AddProperty(global, name, console, DONT_ENUM);
```
AFAICT the property is non enumerable so `console.log(sandbox);` should not _show_ it. It's also configurable so we could delete `global.console`

V8 6.1 in Chrome Canary:
![image](https://user-images.githubusercontent.com/96947/28267931-495f4526-6aca-11e7-8f92-8f6f2d55371b.png)

TimothyGu commented:
Okay, the enumerable bit notwithstanding, `sandbox` might already have a `console` property defined in a different way. It could be an accessor property, for example, and setting it in any way could wreck hazard because of the side effects of the setter. Plus, one would have to save the property descriptor before going into V8 to actually restore to the correct value, and even though that's still not too complex it would be exponentially so if the `console` property is on the [[Prototype]] of the `sandbox` object, or if `sandbox` is a Proxy, etc.

refack commented:
> Okay, the enumerable bit notwithstanding, sandbox might already have a console property defined in a different way. It could be an accessor property, for example, and setting it in any way could wreck hazard because of the side effects of the setter. Plus, one would have to save the property descriptor before going into V8 to actually restore to the correct value, and even though that's still not too complex it would be exponentially so if the console property is on the [[Prototype]] of the sandbox object, or if sandbox is a Proxy, etc.

Ack. Having the ability to flag V8 not to add it, is best.

hashseed commented:
Deleting the console property after creating the context sounds like the simplest solution. You could also get it removed in the snapshot by deleting it before creating the snapshot, but then it will be missing in the main context. That could be solved by actually serializing two separate contexts, but that adds to the snapshot size...

V8 includes the console object as builtin precisely to be able to take advantage of the snapshot.

jasnell commented on doc/api/vm.md:
```diff
@@ -94,12 +99,28 @@ changes:
+(if support for inspector is available in the Node.js build). If not, and if
+`doNotInformInspector` is not specified, this function attaches it to inspector
+using [`inspector.attachContext()`][], and detaches it before returning.
+However, *it is still recommended that you handle context attachment and
```
Please avoid the use of `you` in the docs :-)

jasnell commented on doc/api/vm.md:
```diff
@@ -284,12 +317,28 @@ Returns `true` if the given `sandbox` object has been [contextified][] using
+(if support for inspector is available in the Node.js build). If not, and if
+`doNotInformInspector` is not specified, this function attaches it to inspector
+using [`inspector.attachContext()`][], and detaches it before returning.
+However, *it is still recommended that you handle context attachment and
```
Here too... please avoid `you`

refack commented:
@hashseed what would be the implications of adding:
```patch
   Handle<JSObject> console = factory->NewJSObject(cons, TENURED);
   DCHECK(console->IsJSObject());
 + if (!global.has(name)) {
   JSObject::AddProperty(global, name, console, DONT_ENUM);
 + }
```

hashseed commented:
It doesn't work. We only create one context to capture snapshot from, which acts as template for all contexts.

The sandboxed object is not even copied into the global object, but hooked into it via interceptor, right?

We do have a step where we copy over all properties from the deserialized global object into the global object created from object template. Maybe we can hook that into that place. Iirc the interceptor is disabled during bootstrapping though?

@fhinkel

jgoz commented on doc/api/vm.md:
```diff
@@ -94,12 +99,28 @@ changes:
 
+By default, this function checks if inspector was already aware of the context
+(if support for inspector is available in the Node.js build). If not, and if
+`doNotInformInspector` is not specified, this function attaches it to inspector
```
Lots of negatives in here, which might get confusing. How about something like this:

> This function will attach the context to inspector (in Node.js builds with inspector support) using [`inspector.attachContext()`][] before running the script and detach it with [`inspector.detachContext()`][] before returning. The context will only be attached if inspector was not already aware of the context, or if `doNotInformInspector` is unspecified or `false`. Note that *it is recommended to handle context attachment and detachment manually*, since:

jgoz commented on doc/api/vm.md:
```diff
@@ -284,12 +317,28 @@ Returns `true` if the given `sandbox` object has been [contextified][] using
 
+By default, this function checks if inspector was already aware of the context
+(if support for inspector is available in the Node.js build). If not, and if
+`doNotInformInspector` is not specified, this function attaches it to inspector
```
Same recommended rewording as my above comment.

jgoz commented:
@TimothyGu Given the caveats you outlined in the documentation for automatic attachment/detachment in `vm.runInContext`, would it make more sense for this behaviour to be opt-in, rather than opt-out? I realize that doesn't offer an optimal/"just works" debugging experience, but (especially with the async caveat) the current opt-out behaviour is not optimal either.

eugeneo commented on src/inspector_agent.h:
```diff
@@ -102,6 +137,7 @@ class Agent {
   std::unique_ptr<NodeInspectorClient> client_;
   std::unique_ptr<InspectorIo> io_;
   v8::Platform* platform_;
+  std::vector<const node::inspector::ContextInfo*> contexts_;
```
Please move this to NodeInspectorClient

eugeneo commented on src/inspector_socket_server.cc:
```diff
@@ -566,8 +568,11 @@ void SocketSession::ReadCallback(uv_stream_t* stream, ssize_t read,
-    session->server_->MessageReceived(session->id_,
-                                      std::string(buf->base, read));
+    std::string str(buf->base, read);
+#if DEBUG_TRANSACTION
```
Please remove this before committing.

ak239 commented:
I think we need to fix it on inspector side and allow V8 embedders not to call contextDestroyed and then we can just call contextCreated when context is created in Node.js.
CL: https://chromium-review.googlesource.com/c/575519

TimothyGu commented:
@eugeneo, if you guys can fix it on the V8 side -- even better! We still need some sort of API to provide context name (since `vm Module Context 56` isn't exactly clear on what it actually is) and origin info (for use cases like jsdom if they want to have first class support in inspector; /cc @domenic), though that can be done in `vm.createContext()`.

@hashseed Right now it is actually both [copied](https://github.com/nodejs/node/blob/aa1c8c00437ae2bf5aa0bc980b027013b8dd7e39/src/node_contextify.cc#L118-L138) and intercepted. Because the copying is done synchronously, async in VM contexts is already not supported perfectly so I'd be fine with leaving it that way (/cc @jgoz). In fact, this copying step has led to [a lot of problems](https://github.com/nodejs/node/issues/6283). @AnnaMag is doing some work to rectify that situation in https://github.com/nodejs/node/pull/13265. And yes, the interceptors are [supposed to be](https://github.com/nodejs/node/blob/9ab63d61f8f8d65b0e0643c83dd7936c4cf0a309/src/node_contextify.cc#L386-L388) disabled during bootstrapping.

AnnaMag commented:
@hashseed, @TimothyGu Sandbox object is intercepted, while the copying step is a filler only for the own properties (in the global scope) that were not placed on the sandbox via the interceptors.

eugeneo commented:
I would appreciate if you checked out https://github.com/eugeneo/node/tree/multicontext-v8-patch, as an alternative to that patch.

That branch includes a pending V8 change backported to the Node. I am testing right now, looks like it mostly works - e.g. no crashes and _some_ context are even getting garbage collected... Remaining issues are:
1. Console is always injected into new contexts (this is V8 behavior, not sure what could be done on the Node side).
2. Contexts are retained if "console.log" or "debugger" statements are executed.

I am planning to start a new PR if we manage to resolve issue #2 and if the V8 changes get landed into that tree.

TimothyGu commented:
@eugeneo I'd be happy to defer to your branch if the V8 issues are resolved.

TimothyGu commented:
Closed in favor of #14465.
